cryostatio/cryostat
# fix(jvmid): only replace cached entries if changed

- allow discovery ping period to be set
- fix(jvmid): only replace cached entries if changed

# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
_______________________________________________

Fixes: #1278

## Description of the change:
Adds a check after JVM ID computation to only replace the cached entry if the computed ID has changed.

## Motivation for the change:
The cache emits an event when an entry is updated, even if the value associated with the given key is equal to the previous value.
This results in other internal components triggering on the "update". When the cache is updated because a discovery plugin
re-registers, the target applications published by the plugin are refreshed by the ID cache. The false-positive cache update
notifications cause the targets to be considered lost and immediately refound, which clears other state about the known active
recordings present on the targets. Avoiding the false-positive cache update allows the state to be preserved.

## How to manually test:
1. Run `CRYOSTAT_DISCOVERY_PING_PERIOD=15000 sh smoketest.sh...` from before this PR
2. Select one of the `quarkus-test` applications and manually start a recording. Verify that the default expected recording labels are present
3. Select a different target application (or no selection) and wait 15+ seconds
4. Re-select the original `quarkus-test` application and observe that the recording is still present, but the labels are empty
5. Run `CRYOSTAT_DISCOVERY_PING_PERIOD=15000 CRYOSTAT_IMAGE=ghcr.io/... sh smoketest.sh` with this PR's test image
6. Repeat steps 2-4. Verify that the recording labels are preserved.
